test/should_pass/refine_comparison.erl

```diff
@@ -0,0 +1,26 @@
+-module(refine_comparison).
+
+-export([comp1/1, comp2/1, comp3/1, compatom1/1]).
+
+-spec comp1(integer()) -> pos_integer() | neg_integer().
+comp1(N) when N > 0 -> N;
+comp1(N) when N < 0 -> N;
+comp1(_) -> 1.
+
+-spec comp2(2..10) -> 2..5 | 8..10.
+comp2(N) when N =< 5 -> N;
+comp2(N) when N >= 8 -> N;
+comp2(_) -> 10.
+
+-spec comp3(1..3) -> 1 | 3.
+comp4(N) when N /= 2 -> N;
+comp4(_) -> 1.
+
+-spec compatom1(a | b | c) -> a | b | ok.
+compatom1(X) when X == a -> X;
+compatom1(X) when X =:= b -> X;
+compatom1(_) -> ok.
+
+-spec compatom2(a | b | c | pid()) -> a | c | pid().
+compatom1(X) when X =/= b -> X;
+compatom1(_) -> a.
```

Comment:
I suggest adding the faculty function here as well, written using guards, rather than pattern matching.

```
-spec fac(non_neg_integer()) -> pos_integer().
fac(N) when N == 0 ->
    1;
fac(N) ->
    N * fac(N-1).
```

Reply:
Good idea. This PR doesn't solve this though. The refinement done here only concerns the matching clause. Excluding 0 in subsequent clauses is another PR.